### PR TITLE
Network Messages to Queue (Infra for Multi duties support)

### PR DESCRIPTION
**Pull request by nivBlox**
Move all the network messages to use msgQueue with improved indexes. Also identifier is now made from publicKey & duty role for better indexing. 

* Infra for multi duties (aggregate & propose)

### CODE REVIEW

File: beacon/beacon.go [hunk 1]
```diff
@@ -50,19 +49,19 @@ type Beacon interface {
 	SubmitAttestation(ctx context.Context, attestation *ethpb.Attestation, validatorIndex uint64, key *bls.PublicKey) error
 
 	// GetAggregationData returns aggregation data for the given slot and committee index
-	GetAggregationData(ctx context.Context, duty slotqueue.Duty) (*ethpb.AggregateAttestationAndProof, error)
+	GetAggregationData(ctx context.Context, duty *ethpb.DutiesResponse_Duty, key *bls.PublicKey, shareKey *bls.SecretKey) (*ethpb.AggregateAttestationAndProof, error)
 
 	// SignAggregation signs the given aggregation data
-	SignAggregation(ctx context.Context, data *ethpb.AggregateAttestationAndProof, duty slotqueue.Duty) (*ethpb.SignedAggregateAttestationAndProof, error)
+	SignAggregation(ctx context.Context, data *ethpb.AggregateAttestationAndProof, duty *bls.SecretKey) (*ethpb.SignedAggregateAttestationAndProof, error)
```
Comment (amir-blox) [inline beacon/beacon.go:55]: change var name (`duty`)

File: beacon/prysmgrpc/aggrate.go [hunk 2]
```diff
@@ -68,18 +69,20 @@ func (b *prysmGRPC) SubmitAggregation(ctx context.Context, data *ethpb.SignedAgg
...
 
+	slotSig, err := b.signSlot(ctx, slot, shareKey)
+	if err != nil {
+		return false, err
+	}
+
 	hash := hashutil.Hash(slotSig)
 	val := binary.LittleEndian.Uint64(hash[:8])%modulo == 0
+
+	b.logger.Info("check if is aggregator", zap.Int("committee", committeeLen), zap.Uint64("slot", slot), zap.Any("hash little endian", binary.LittleEndian.Uint64(hash[:8])), zap.Uint64("modulo", binary.LittleEndian.Uint64(hash[:8])%modulo))
```
Comment (amir-blox) [inline beacon/prysmgrpc/aggrate.go:85]: seems like a `Debug` log

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -32,7 +32,7 @@ Build stage Docker image:
     - docker tag $IMAGE_NAME:$CI_BUILD_REF $DOCKER_REPO_INFRA_STAGE:$CI_BUILD_REF
     - $DOCKER_LOGIN_TO_INFRA_STAGE_REPO && docker push $DOCKER_REPO_INFRA_STAGE:$CI_BUILD_REF
   only:
-    - stage
+    - multi-duties
```
Comment (alonmuroch) [inline .gitlab-ci.yml]: @nivBlox should we change it?

File: beacon/beacon.go [hunk 4]
```diff
@@ -50,19 +49,19 @@ type Beacon interface {
 	SubmitAttestation(ctx context.Context, attestation *ethpb.Attestation, validatorIndex uint64, key *bls.PublicKey) error
 
 	// GetAggregationData returns aggregation data for the given slot and committee index
-	GetAggregationData(ctx context.Context, duty slotqueue.Duty) (*ethpb.AggregateAttestationAndProof, error)
+	GetAggregationData(ctx context.Context, duty *ethpb.DutiesResponse_Duty, key *bls.PublicKey, shareKey *bls.SecretKey) (*ethpb.AggregateAttestationAndProof, error)
```
Comment (alonmuroch) [inline beacon/beacon.go:52]: @nivBlox I think we should remove this for this PR

File: beacon/prysmgrpc/aggrate.go [hunk 5]
```diff
@@ -20,18 +20,19 @@ import (
...
+	slotSig, err := b.signSlot(ctx, duty.AttesterSlot, shareKey)
 	if err != nil {
 		return nil, errors.Wrap(err, "could not sign slot")
 	}
 
 	res, err := b.validatorClient.SubmitAggregateSelectionProof(ctx, &ethpb.AggregateSelectionRequest{
-		Slot:           duty.Duty.AttesterSlot,
-		CommitteeIndex: duty.Duty.CommitteeIndex,
-		PublicKey:      duty.ValidatorPK.Serialize(),
+		Slot:           duty.AttesterSlot,
```
Comment (alonmuroch) [inline beacon/prysmgrpc/aggrate.go:33]: @nivBlox let's remove this from this PR and move it for the actual proposal PR

File: ibft/ibft.go [hunk 6]
```diff
@@ -85,6 +87,8 @@ func New(
 }
 
 func (i *ibftImpl) Init() {
+	i.listenToDecidedQueueMessages()
+	i.listenToSyncQueueMessages(i.ValidatorShare.PublicKey.Serialize()) // pass the pubkey byte to avoid deadlock
```
Comment (alonmuroch) [inline ibft/ibft.go:91]: @nivBlox we have the identifier set in the sturct... why pass another similar argument here?

File: ibft/ibft.go [hunk 7]
```diff
@@ -105,7 +109,7 @@ func (i *ibftImpl) StartInstance(opts StartOptions) (bool, int, []byte) {
 	go newInstance.StartMessagePipeline()
 	stageChan := newInstance.GetStageChan()
 
-	err := i.resetLeaderSelection(opts.Identifier) // Important for deterministic leader selection
+	err := i.resetLeaderSelection(i.Identifier) // Important for deterministic leader selection
```
Comment (alonmuroch) [inline ibft/ibft.go:112]: @nivBlox we don't need to pass a struct variable to a struct function

File: ibft/ibft.go [hunk 8]
```diff
@@ -148,7 +152,17 @@ func (i *ibftImpl) GetIBFTCommittee() map[uint64]*proto.Node {
...
+	return i.Identifier
+}
+
 // resetLeaderSelection resets leader selection with seed and round 1
 func (i *ibftImpl) resetLeaderSelection(seed []byte) error {
 	return i.leaderSelector.SetSeed(seed, 1)
 }
+
+// IdentifierFormat return base format for lambda
+func IdentifierFormat(pubKey []byte, role beacon.Role) string {
```
Comment (alonmuroch) [inline ibft/ibft.go:166]: @nivBlox are we using it only in the ibft implementation? 
Feels like this should be in some util function

File: ibft/ibft_decided.go [hunk 9]
```diff
@@ -4,17 +4,32 @@ import (
...
+	ibft_sync "github.com/bloxapp/ssv/ibft/sync"
+	"github.com/bloxapp/ssv/network/msgqueue"
 	"github.com/bloxapp/ssv/storage/collections"
 	"github.com/pkg/errors"
 	"go.uber.org/zap"
+	"time"
 )
 
+// listenToDecidedQueueMessages is listen for all the ibft decided msg's and process them
+func (i *ibftImpl) listenToDecidedQueueMessages() {
```
Comment (alonmuroch) [inline ibft/ibft_decided.go:16]: @nivBlox naming is wrong.. should be ProcessDecidedMessages

File: ibft/pipeline/auth/msg_lambda.go [hunk 10]
```diff
@@ -18,3 +19,17 @@ func ValidateLambdas(lambda []byte) pipeline.Pipeline {
 		return nil
 	})
 }
+
+// ValidateLambdasRole validates msg role with ibft controller role
+func ValidateLambdasRole(role string) pipeline.Pipeline {
```
Comment (alonmuroch) [inline ibft/pipeline/auth/msg_lambda.go:24]: @nivBlox iBFT doesn't know what role is.. let's keep it iBFT generic.. this should be ValidateLambdas

File: ibft/ibft_decided.go [hunk 11]
```diff
@@ -60,7 +75,7 @@ func (i *ibftImpl) ProcessDecidedMessage(msg *proto.SignedMessage) {
 			i.currentInstance.Stop()
 		}
 		// sync
-		s := sync.NewHistorySync(i.logger, msg.Message.ValidatorPk, i.network, i.ibftStorage, i.validateDecidedMsg)
+		s := ibft_sync.NewHistorySync(i.logger, msg.Message.ValidatorPk, i.network, i.ibftStorage, i.validateDecidedMsg)
```
Comment (alonmuroch) [inline ibft/ibft_decided.go:78]: @nivBlox why call it ibft_sync?

File: ibft/ibft_decided_test.go [hunk 12]
```diff
@@ -270,7 +271,7 @@ func TestSyncAfterDecided(t *testing.T) {
 		Round:       3,
 		SeqNumber:   10,
 		ValidatorPk: validatorPK(sks).Serialize(),
-		Lambda:      []byte("Lambda"),
+		Lambda:      []byte(IdentifierFormat(validatorPK(sks).Serialize(), beacon.RoleAttester)),
```
Comment (alonmuroch) [inline ibft/ibft_decided_test.go:274]: @nivBlox test is generic.. it doesn't matter what the lambda is.. let's change it back

File: ibft/ibft_network.go [hunk 13]
```diff
@@ -43,9 +42,11 @@ func (i *ibftImpl) listenToNetworkMessages() {
 	decidedChan := i.network.ReceivedDecidedChan()
 	go func() {
 		for msg := range decidedChan {
-			if bytes.Equal(i.ValidatorShare.PublicKey.Serialize(), msg.Message.ValidatorPk) { // making sure the msg is relevant to the share only
```
Comment (alonmuroch) [inline ibft/ibft_network.go:46]: @nivBlox why did we remove it? because identifier has public key in it?

File: ibft/ibft_sync.go [hunk 14]
```diff
@@ -1,19 +1,36 @@
...
-	"github.com/bloxapp/ssv/ibft/sync"
+	ibft_sync "github.com/bloxapp/ssv/ibft/sync"
 	"github.com/bloxapp/ssv/network"
+	"github.com/bloxapp/ssv/network/msgqueue"
 	"go.uber.org/zap"
+	"time"
 )
 
+// listenToSyncQueueMessages is listen for all the ibft sync msg's and process them
+func (i *ibftImpl) listenToSyncQueueMessages(pubKey []byte) {
```
Comment (alonmuroch) [inline ibft/ibft_sync.go:12]: @nivBlox naming is wrong .. should be ProcessSyncMessages

File: ibft/ibft_sync.go [hunk 15]
```diff
@@ -1,19 +1,36 @@
...
+	go func() {
+		for {
+			if syncMsg := i.msgQueue.PopMessage(msgqueue.SyncIndexKey(pubKey)); syncMsg != nil {
```
Comment (alonmuroch) [inline ibft/ibft_sync.go:15]: @nivBlox we shouldn't pass pub key but rather identifier

File: ibft/ibft_sync_test.go [hunk 16]
```diff
@@ -43,7 +46,8 @@ func aggregateSign(t *testing.T, sks map[uint64]*bls.SecretKey, msg *proto.Messa
 func populatedStorage(t *testing.T, sks map[uint64]*bls.SecretKey, highestSeq int) collections.Iibft {
 	storage := collections.NewIbft(newInMemDb(), zap.L(), "attestation")
 	for i := 0; i <= highestSeq; i++ {
-		lambda := []byte(fmt.Sprintf("lambda_%d", i))
```
Comment (alonmuroch) [inline ibft/ibft_sync_test.go:46]: @nivBlox lambda should be generic.. let's change it back

File: ibft/ibft_sync_test.go [hunk 17]
```diff
@@ -140,6 +139,7 @@ func TestSyncFromScratch100Sequences(t *testing.T) {
 	_ = populatedIbft(2, network, populatedStorage(t, sks, 100), sks, nodes)
 
 	i1.(*ibftImpl).SyncIBFT()
+	time.Sleep(time.Millisecond * 500) // wait for sync to complete
```
Comment (alonmuroch) [inline ibft/ibft_sync_test.go:142]: @nivBlox not sure what changed that we need to add sleeps in between?

File: network/msgqueue/indexes.go [hunk 18]
```diff
@@ -7,31 +7,60 @@ import (
...
+				DecidedIndexKey(msg.Lambda),
+			}
+		}
+		return []string{}
+	}
+}
+
+// SyncIndexKey is the ibft sync index key
+func SyncIndexKey(pubKey []byte) string {
+	return fmt.Sprintf("sync_publicKey_%s", hex.EncodeToString(pubKey))
```
Comment (alonmuroch) [inline network/msgqueue/indexes.go:57]: @nivBlox should receive identifier not public key

File: node/valcheck/aggregator.go [hunk 19]
```diff
@@ -0,0 +1,27 @@
+package valcheck
+
```
Comment (alonmuroch) [inline node/valcheck/aggregator.go:2]: @nivBlox let's remove it from this PR and move it to when we have the aggregator PR

File: node/valcheck/proposer.go [hunk 20]
```diff
@@ -0,0 +1,27 @@
+package valcheck
```
Comment (alonmuroch) [inline node/valcheck/proposer.go:1]: @nivBlox let's remove it from this PR and move it to the PR containing proposals

File: validator/duty_execution.go [hunk 21]
```diff
@@ -136,63 +137,66 @@ func (v *Validator) comeToConsensusOnInputValue(ctx context.Context, logger *zap
...
-	//		return 0, nil, errors.Wrap(err, "failed to get proposal block")
-	//	}
-	//
-	//	inputValue.Data = &proto.InputValue_BeaconBlock{
-	//		BeaconBlock: block,
-	//	}
-	default:
-		return 0, nil, nil, errors.Errorf("unknown role: %s", role.String())
-	}
+	case beacon.RoleAggregator:
```
Comment (alonmuroch) [inline validator/duty_execution.go:143]: @nivBlox let's remove aggregator and proposal roles for now

File: validator/signature.go [hunk 22]
```diff
@@ -45,29 +45,48 @@ func (v *Validator) signDuty(ctx context.Context, decidedValue []byte, role beac
...
-	//	err = e
-	//	sig = signedAggregation.GetSignature()
-	//case beacon.RoleProposer:
-	//	signedProposal, e := v.beacon.SignProposal(ctx, inputValue.GetBeaconBlock())
-	//	inputValue.SignedData = &proto.InputValue_Block{
-	//		Block: signedProposal,
-	//	}
-	//	err = e
-	//	sig = signedProposal.GetSignature()
+	case beacon.RoleAggregator:
```
Comment (alonmuroch) [inline validator/signature.go:60]: @nivBlox let's remove aggregator and proposal duties for now

**Review by alonmuroch — CHANGES_REQUESTED** [review]

File: .gitlab-ci.yml (hunk 3 above)
Comment (nivBlox) [reply .gitlab-ci.yml]: @alonmuroch changed before merge

File: beacon/beacon.go (hunk 4 above)
Comment (nivBlox) [reply beacon/beacon.go:52]: We are not supporting aggs & propose in this merge. no ibft's for those

File: beacon/prysmgrpc/aggrate.go (hunk 5 above)
Comment (nivBlox) [reply beacon/prysmgrpc/aggrate.go:33]: in my opinion its better to checkout from this branch with this code inside to reduce additional effort

File: ibft/ibft.go (hunk 6 above)
Comment (nivBlox) [reply ibft/ibft.go:91]: cause sync is indexing only with pubkey (role is not relevant for this type of msg's)

File: network/msgqueue/indexes.go (hunk 18 above)
Comment (nivBlox) [reply network/msgqueue/indexes.go:57]: identifier is built from pubkey and role. sync msg's not familiar with role

Comment (alonmuroch) [reply network/msgqueue/indexes.go:57]: every message has an identifier.. we either identify msgs by identifier or something else but let's not mix

File: ibft/ibft.go (hunk 7 above)
Comment (nivBlox) [reply ibft/ibft.go:112]: the func is excepting "seed" so it seems that it should remain generic

File: network/msgqueue/indexes.go (hunk 18 above)
Comment (nivBlox) [reply network/msgqueue/indexes.go:57]: each msg type has differ index you cant force the same for all.. 
also sync is at validator level that doesnt have identifier

File: node/valcheck/aggregator.go (hunk 19 above)
Comment (nivBlox) [reply node/valcheck/aggregator.go:2]: i sugget to leave it for the next pr to save us to write the same code
@lior-blox

File: node/valcheck/proposer.go (hunk 20 above)
Comment (nivBlox) [reply node/valcheck/proposer.go:1]: i sugget to leave it for the next pr to save us to write the same code
@lior-blox

File: validator/duty_execution.go (hunk 21 above)
Comment (nivBlox) [reply validator/duty_execution.go:143]: i sugget to leave it for the next pr to save us to write the same code
@lior-blox

File: validator/signature.go (hunk 22 above)
Comment (nivBlox) [reply validator/signature.go:60]: i sugget to leave it for the next pr to save us to write the same code
@lior-blox

File: ibft/ibft.go (hunk 7 above)
Comment (alonmuroch) [reply ibft/ibft.go:112]: @nivBlox we should change resetLeaderSelection to use identifier + seq number

**Pull request merged**